packages/gatsby/src/commands/build-html.ts

```diff
@@ -46,20 +45,83 @@ interface IBuildRendererResult {
 }
 
 let devssrWebpackCompiler: webpack.Watching | undefined
-let needToRecompileSSRBundle = true
+let SSRBundleReceivedInvalidEvent = true
+let SSRBundleWillInvalidate = false
+
+export function devSSRWillInvalidate(): void {
+  // we only get one `invalid` callback, so if we already
+  // set this to true, we can't really await next `invalid` event
+  if (!SSRBundleReceivedInvalidEvent) {
+    SSRBundleWillInvalidate = true
+  }
+}
+
+let activeRecompilations = 0
+
 export const getDevSSRWebpack = (): {
-  devssrWebpackWatcher: IWebpackWatchingPauseResume
-  devssrWebpackCompiler: webpack.Compiler
+  recompileAndResumeWatching: (
+    allowTimedFallback: boolean
+  ) => Promise<() => void>
   needToRecompileSSRBundle: boolean
 } => {
   if (process.env.gatsby_executing_command !== `develop`) {
     throw new Error(`This function can only be called in development`)
   }
 
-  return {
-    devssrWebpackWatcher: devssrWebpackCompiler as webpack.Watching,
-    devssrWebpackCompiler: (devssrWebpackCompiler as webpack.Watching).compiler,
-    needToRecompileSSRBundle,
+  const watcher = devssrWebpackCompiler as webpack.Watching
+  const compiler = (devssrWebpackCompiler as webpack.Watching).compiler
+  if (watcher && compiler) {
+    return {
+      recompileAndResumeWatching: async function recompileAndResumeWatching(
+        allowTimedFallback: boolean
+      ): Promise<() => void> {
+        let isResolved = false
+        return await new Promise<() => void>(resolve => {
+          function stopWatching(): void {
+            activeRecompilations--
+            if (activeRecompilations === 0) {
+              watcher.suspend()
+            }
+          }
+          let timeout
+          function finish(): void {
+            emitter.off(`DEV_SSR_COMPILATION_DONE`, finish)
+            if (!isResolved) {
+              isResolved = true
+              resolve(stopWatching)
+            }
+            if (timeout) {
+              clearTimeout(timeout)
+            }
+          }
+          emitter.on(`DEV_SSR_COMPILATION_DONE`, finish)
+          // we reset it before we start compilation to be able to catch any invalid events
+          SSRBundleReceivedInvalidEvent = false
+          if (activeRecompilations === 0) {
+            watcher.resume()
+          }
+          activeRecompilations++
+
+          if (allowTimedFallback) {
+            // Timeout after 1.5s.
```

Comment:
I know we had this the same previously (and it's probably arbitrary). Any good reasons to keep it as-is or increase/decrease it? I'm fine with keeping it like that

Reply:
I decided not to change any of timeouts we already have in this PR - the goal was only to make it work again at all here, and as I mentioned couple of time - there is a lot of opporunities to improve the way DEV_SSR works, I just didn't want this PR to get stuck on those questions, as improvements should be a follow up to changes done here which are mostly fixes